### PR TITLE
Add ut coverage for capabilities.Setup

### PR DESCRIPTION
**Pull request by robert-cronin**
#### What type of PR is this?

/kind cleanup
#### What this PR does / why we need it:

Increases Unit Test coverage for `pkg/capabilities` to 100% by adding test cases for `capabilities.Setup`

#### Which issue(s) this PR fixes:

N/A

#### Special notes for your reviewer:

This pull request introduces unit test coverage for the `Setup` function. Although there is an existing method, `SetForTests`, designed to facilitate testing, there has been no coverage for the actual `Setup` method until now.

If I've missed any guidelines or made any mistakes, I apologize and welcome your feedback.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

File: pkg/capabilities/capabilities_test.go [hunk 1]
```diff
@@ -55,3 +55,58 @@ func TestGet(t *testing.T) {
...
+
+// compareCapabilities compares two Capabilities instances
+func compareCapabilities(a, b Capabilities) bool {
+	if a.AllowPrivileged != b.AllowPrivileged {
+		return false
+	}
+	if a.PerConnectionBandwidthLimitBytesPerSec != b.PerConnectionBandwidthLimitBytesPerSec {
+		return false
+	}
+	return true
```
Comment (bart0sh) [inline pkg/capabilities/capabilities_test.go:111]:
```suggestion
	return a.PerConnectionBandwidthLimitBytesPerSec == b.PerConnectionBandwidthLimitBytesPerSec
```

Comment (robert-cronin) [reply pkg/capabilities/capabilities_test.go]: Updated the function to include the suggested comparison, thanks for the suggestion!

File: pkg/capabilities/capabilities_test.go [hunk 2]
```diff
@@ -55,3 +55,55 @@ func TestGet(t *testing.T) {
...
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			defer func() {
+				capInstance.lock.Lock()
+				defer capInstance.lock.Unlock()
+				capInstance.capabilities = nil
+				capInstance.once = sync.Once{}
+			}()
```
Comment (bart0sh) [inline pkg/capabilities/capabilities_test.go:92]: Would it make sense to add `capInstance.once = sync.Once{}` to the `SetForTests` and use it in this and possibly other test cases?

Comment (robert-cronin) [reply pkg/capabilities/capabilities_test.go:92]: Hi @bart0sh, thank you for your suggestion. I agree with your line of thinking; however, I think using SetForTests here would bypass Setup, defeating the purpose of testing how Setup actually initializes capabilities.

Am I misunderstanding what you mean?

Comment (bart0sh) [reply pkg/capabilities/capabilities_test.go:92]: I thought that this code is almost the same as `SetForTests`, so you can change `SetForTests`:
```
 // SetForTests sets capabilities for tests.  Convenience method for testing.  This should only be called from tests.
-func SetForTests(c Capabilities) {
+func SetForTests(c *Capabilities) {
        capInstance.lock.Lock()
        defer capInstance.lock.Unlock()
-       capInstance.capabilities = &c
+       capInstance.capabilities = c
+       capInstance.once = sync.Once{}
 }
```

and then use it in the tests like this:
```
        for _, tc := range testCases {
                t.Run(tc.name, func(t *testing.T) {
-                       defer func() {
-                               capInstance.lock.Lock()
-                               defer capInstance.lock.Unlock()
-                               capInstance.capabilities = nil
-                               capInstance.once = sync.Once{}
-                       }()
+                       defer SetForTests(nil)
```

Changing the name of SetForTests would probably make sense too.

WDYT?

Comment (robert-cronin) [reply pkg/capabilities/capabilities_test.go:92]: Thank you for the clarification, @bart0sh. I see what you mean now. You're suggesting we modify the SetForTests function itself and use it to reset the capabilities for each test. I think this is a good approach.

Also, I agree that changing the name would make sense. How about `ResetCapabilitiesForTest`? This seems more descriptive of the function's purpose. I'll include the rename in the upcoming changes :+1:

Comment (bart0sh) [reply pkg/capabilities/capabilities_test.go]: I'd suggest to name it `ResetForTest` as `capabilities.SetCapabilitiesForTest` is long and repeats `capabilities` 2 times.

Comment (robert-cronin) [reply pkg/capabilities/capabilities_test.go]: I agree, `ResetForTest` is clearer and more concise.

Comment (bart0sh) [reply pkg/capabilities/capabilities_test.go]: Thank you for the quick update!
/lgtm

Comment (bart0sh) [reply pkg/capabilities/capabilities_test.go]: one more thing: This code should use `ResetForTest` as well, I believe:
```
func TestGet(t *testing.T) {
	defer func() {
		capInstance.lock.Lock()
		defer capInstance.lock.Unlock()
		capInstance.capabilities = nil
		capInstance.once = sync.Once{}
	}()
```

Comment (bart0sh) [reply pkg/capabilities/capabilities_test.go]: or should be removed as `ResetForTest` is already called in this test case.

Comment (robert-cronin) [reply pkg/capabilities/capabilities_test.go]: Oh yes I missed that test case, thank you for catching that! It has been updated in the latest changes

File: pkg/capabilities/capabilities_test.go [hunk 3]
```diff
@@ -48,10 +42,57 @@ func TestGet(t *testing.T) {
...
+			expectedCapabilities: Capabilities{
+				AllowPrivileged:                        false,
+				PerConnectionBandwidthLimitBytesPerSec: 2048,
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			defer ResetForTest(nil)
```
Comment (bart0sh) [inline pkg/capabilities/capabilities_test.go:81]: Just wondering if this call should be moved up to avoid resetting 2 times?

Comment (robert-cronin) [reply pkg/capabilities/capabilities_test.go:81]: I might be misunderstanding what you mean, but I believe it's necessary to keep the `ResetForTest(nil)` call inside each test case to ensure each test starts with a clean state. This is confirmed by the fact that if I take it out, the second test case fails due to the `Once` not being reset properly and the subsequent call to `capabilities.once.Do` being ignored.

Comment (bart0sh) [reply pkg/capabilities/capabilities_test.go:81]: My bad. I thought that `ResetForTest` will be called after the loop. I didn't notice that it's called in the function.

Comment (bart0sh) [inline pkg/capabilities/capabilities_test.go:81]: BTW, would it make sense to call it directly before the test case like this?
```suggestion
			ResetForTest(nil)
```

File: pkg/capabilities/capabilities_test.go [hunk 4]
```diff
@@ -48,10 +42,57 @@ func TestGet(t *testing.T) {
 			HostNetworkSources: []string{"A", "B"},
 		},
 	}
-	SetForTests(cap)
+	ResetForTest(&cap)
```
Comment (bart0sh) [inline pkg/capabilities/capabilities_test.go:45]: Here we call `ResetForTest` twice - before the test and after it. I think calling it after the test may be redundant.

File: pkg/capabilities/capabilities_test.go (hunk 3 above)
Comment (robert-cronin) [reply pkg/capabilities/capabilities_test.go:81]: I agree, resetting it before the test makes more sense

File: pkg/capabilities/capabilities_test.go (hunk 4 above)
Comment (robert-cronin) [reply pkg/capabilities/capabilities_test.go:45]: You're right, the call to ResetForTest after the test is indeed redundant, I'll take it out!

Comment (bart0sh) [reply pkg/capabilities/capabilities_test.go:44]: /lgtm

**Comment by bart0sh** [comment]
/assign @wojtek-t @dchen1107 @dims @liggitt @smarterclayton
for approval

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm
/approve

File: pkg/capabilities/capabilities.go [hunk 5]
```diff
@@ -68,11 +68,13 @@ func Setup(allowPrivileged bool, perConnectionBytesPerSec int64) {
...
-// SetForTests sets capabilities for tests.  Convenience method for testing.  This should only be called from tests.
-func SetForTests(c Capabilities) {
+// ResetForTest resets the capabilities to a given state for testing purposes.
+// This function should only be called from tests.
+func ResetForTest(c *Capabilities) {
 	capInstance.lock.Lock()
 	defer capInstance.lock.Unlock()
-	capInstance.capabilities = &c
+	capInstance.capabilities = c
+	capInstance.once = sync.Once{}
```
Comment (liggitt) [inline pkg/capabilities/capabilities.go:77]: this function is confusing now, and combines reset AND Initialize/Setup

if we want a reset function, and for all sets to go through the normal code path, don't let this function take a `c *Capabilities` parameter, just make this function reset:

```go
func ResetForTest(c *Capabilities) {
	capInstance.lock.Lock()
	defer capInstance.lock.Unlock()
	capInstance.capabilities = nil
	capInstance.once = sync.Once{}
```

Comment (robert-cronin) [reply pkg/capabilities/capabilities.go:77]: Hi @liggitt 
I've refactored the method as you suggested, so now it is truly only resetting the capInstance and not doing anything else. For intialize/setup, I am now using the Initialize method so as to keep to the normal code path as much as possible. Thanks for the suggestion!

File: pkg/capabilities/capabilities_test.go [hunk 6]
```diff
@@ -48,10 +41,58 @@ func TestGet(t *testing.T) {
...
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			ResetForTest()
+
+			Setup(tc.allowPrivileged, tc.perConnectionBytesPerSec)
+			res := Get()
+			if !compareCapabilities(tc.expectedCapabilities, res) {
```
Comment (liggitt) [inline pkg/capabilities/capabilities_test.go:85]: should be able to directly compare `tc.expectedCapabilities != res`

File: pkg/capabilities/capabilities_test.go [hunk 7]
```diff
@@ -18,17 +18,10 @@ package capabilities
 
 import (
 	"reflect"
-	"sync"
 	"testing"
 )
 
 func TestGet(t *testing.T) {
-	defer func() {
```
Comment (liggitt) [inline pkg/capabilities/capabilities_test.go:26]: leave this defer in place to leave the package in the same state when this test is done

File: pkg/capabilities/capabilities_test.go (hunk 6 above)
Comment (robert-cronin) [reply pkg/capabilities/capabilities_test.go:85]: I ran into an issue when trying this one: the `PrivilegedSources` field is not a comparable type. For now, I have just changed it to use `reflect.DeepEqual` which I think addresses the comment. Happy to change it to something else.

File: pkg/capabilities/capabilities_test.go (hunk 7 above)
Comment (robert-cronin) [reply pkg/capabilities/capabilities_test.go:26]: done! thanks @liggitt

**Pull request merged**